src/ops.jl

```diff
@@ -82,39 +82,102 @@ end
-                groups)
+             reverse(stride), length(stride),
+             reverse(padding), length(padding),
+             reverse(dilation), length(dilation),
+             groups)
 
   Tensor{T,N}(ptr[], on(input))
 end
 
+function conv_transpose_2d(input::Tensor{T}, filter::Tensor{T,N}, bias::Tensor{T};
+        stride = [1],
+        padding = [0],
+        output_padding = [0],
+        dilation = [1],
+        groups = 1) where {T,N}
+
+  ptr = Ref(Ptr{Cvoid}())
+
+  atg_conv_transpose2d(ptr, input.ptr, filter.ptr, bias.ptr,
+						reverse(stride), length(stride),
+                       reverse(padding), length(padding),
+                       reverse(output_padding), length(output_padding),
+                       groups,
+                       reverse(dilation), length(dilation))
+
+  Tensor{T,N}(ptr[], on(input))
+end
+
+function _depthwise_conv2d(input::Tensor{T}, filter::Tensor{T,N}, bias::Tensor{T};
+                           stride = [1],
+                           padding = [0],
+                           dilation = [1]) where {T,N}
+
+    # When groups == in_channels and out_channels == K * in_channels, where K is a positive integer,
+    # this operation is also termed in literature as depthwise convolution.
+
+    c_in = size(input)[end - 1]  # number of input channels
+    c_out = size(filter)[end]    # number of output channels
+    @assert mod(c_in, c_out) == 0 "Invalid kernel size for depthwise convolution"
+
+    groups = c_in
+    ptr = Ref(Ptr{Cvoid}())
+
+    atg_conv2d(ptr, input.ptr, filter.ptr, bias.ptr,
+               reverse(stride), length(stride),
+               reverse(padding), length(padding),
+               reverse(dilation), length(dilation),
+               groups)
+
+    Tensor{T,N}(ptr[], on(input))
+end
+
 function _softmax(input::Tensor{T,N}, dims = 1, dtype = options[T]) where {T,N}
   ptr = Ref(Ptr{Cvoid}())
 
-  atg_softmax(ptr, input.ptr, N - dims - 1, dtype)
+  atg_softmax(ptr, input.ptr, N - dims, dtype)
   Tensor{T,N}(ptr[], on(input))
 end
 
-function _meanpool(t::Tensor{T,N}, k, s, p, op_sz) where {T,N}
+function _meanpool(t::Tensor{T,N}, kernel_size; stride = [1] , padding = [0]) where {T,N}
+  k = collect(kernel_size)
+  s = collect(stride)
+  p = collect(padding)
   ptr = Ref(Ptr{Cvoid}())
 
   atg_avg_pool2d(ptr, t.ptr,
-                 k, length(k),
-                 s, length(s),
-                 p, length(p),
-                 0,                # ceil_mode
-                 1,                # count_include_pad
-                 1                 # divisor_override
+                 reverse(k), length(k),
+                 reverse(s), length(s),
+                 reverse(p), length(p),
+                 0,  # ceil_mode
+                 1,  # count_include_pad
+                 prod(k)  # divisor_override
+  )
+  Tensor{T,N}(ptr[], on(t))
+end
+
+function _maxpool(t::Tensor{T,N}, kernel_size; stride = [1], padding = [0], dilation = [1]) where {T,N}
```

Comment:
Why do we need this method?

Reply:
I did not have the intention at first to make a PR, so I changed a few things that was not really needed. Sorry for that.

In my mind it is more clean to have the NNlib ConvDims and PoolDims only in nnlib.jl and not in the ops.jl. Also it makes sense to align the arguments of _mean_pool and _max_pool. My own preference would be to remove the _maxpool method with PoolDims.

Reply:
Best to keep the consistency with the method signatures and API for now.

Reply:
I was not aware of the #21 issue. Is this PR still interesting? I can clean it up to include only the tests and the changes needed to pass the tests. In other words, remove the added layers, leave the method signatures as they are. Reversing strides, pads, dilations... to match the spatial sizes in `ops.jl` is however still needed.

I guess this will be resolved by #21 at some point.

Reply:
The PR is definitely interesting. #21 should be fixed in the latest release too. 

Keeping the signatures consistent is important however. The layers themselves should be fine. Reversing the dims is right too, I think.